EssentialsDiscord/src/com/earth2me/essentials/discord/utils/MessageType.java

```diff
@@ -1,49 +0,0 @@
-package com.earth2me.essentials.discord.utils;
-
-/**
- * Types of messages that can be sent to or received from Discord by this plugin.
- * Other plugins may also send messages, but these are not listed hre.
- */
-public enum MessageType {
-
-    GLOBAL_CHAT("chat", true, false),
-    LOCAL_CHAT("local-chat", true, false),
-    HELPOP("helpop", true, false),
-    BROADCAST("broadcast"),
-    PLAYER_JOIN("join"),
-    PLAYER_QUIT("quit"),
-    PLAYER_DEATH("death"),
-    PLAYER_ADVANCEMENT("advancement"),
-    PLAYER_MUTE("mute"),
-    PLAYER_KICK("kick"),
-    PLAYER_BAN("ban"),
-    PLAYER_AWAY("afk"),
-    CONSOLE_LOG("console", false, true),
-    OTHER("other");
-
-    private final String configName;
-    private final boolean canWebhook;
-    private final boolean shouldBatch;
-
-    MessageType(final String configName) {
-        this(configName, false, false);
-    }
-
-    MessageType(final String configName, final boolean canWebhook, final boolean shouldBatch) {
-        this.configName = configName;
-        this.canWebhook = canWebhook;
-        this.shouldBatch = shouldBatch;
-    }
-
-    public String getConfigName() {
-        return configName;
-    }
-
-    public boolean isCanWebhook() {
-        return canWebhook;
-    }
-
-    public boolean isShouldBatch() {
-        return shouldBatch;
-    }
-}
```

Comment:
This enum was added with the intention of being able to have a 1:N mapping of Discord channels to types of messages that can be received. I feel we should keep some sort of abstraction over the types of messages that get sent by this plugin to Discord to avoid magic strings and to make it clearer to third-party plugins how to integrate with the plugin.

Reply:
Yeah so the main reason why I removed it was because it seems quite limiting for the API, since using MessageType would restrict config values to a specific set defined in the enum, whereas plugins might have their own custom message type which could be controlled from the Essentials config by adding a key. The important thing would just be to document what the type string is used for, but otherwise I don't think it would be very unclear. We could re-add the enum for Essentials messages, but then we have to handle for both receiving messages with an enum type and a string type which sounds a bit messy.

Reply:
We could have a `MessageType` interface which exposes these meta options as methods (whether to batch and whether to support sending by webhooks), then have a `StandardTypes` enum for inbuilt messages. It's been a long time since I started this branch (and my commit history is probably gone for good now) but I think that was the plan.

Reply:
Looks like it was added in https://github.com/EssentialsX/Essentials/commit/2f8f969f1641c4723d406c8a21366101ba984798 on #2488 (which hasn't been closed yet, but this PR is basically an updated version of that). I started working on this before realizing it was a copy of the other branch, with commit history nuked.

Reply:
I was just thinking, maybe a builder pattern here would be better for controlling how messages are constructed. This would allow for abstracting over this without necessitating enums everywhere. The builder can have overloaded methods for things like the type, messages (maybe allows supporting chat components later on, etc), and how/if it should batch. The `build()` method can then validate that the message is constructed correctly.

Event contructor
`public DiscordMessageEvent(DiscordMessage message)`

DiscordMessage builder examples
`DiscordMessage.builder().type(MessageType.CHAT).message("Hello").build()`
`DiscordMessage.builder().type("myplugin").message("Test").batch(true).build()`

I was also thinking, maybe the plugin sending the message should also be included when building the message. This would allow us to differentiate messages sent by Essentials vs another plugin (especially when they could use the same type), and allow for some degree of control over how messages from other plugins are processed.

Any thoughts on this?